example/mapboxExampleCamera.js

```diff
@@ -0,0 +1,284 @@
+	else {
+
+		var pixelsPerMeter = projectedUnitsPerMeter( coords[ 1 ] );
+		projected.push( coords[ 2 ] * pixelsPerMeter );
+
+	}
+
+	var result = new THREE.Vector3( projected[ 0 ], projected[ 1 ], projected[ 2 ] );
+
+	return result;
+
+}
+
+function projectedUnitsPerMeter( latitude ) {
+
+	return Math.abs(
+		Constants.WORLD_SIZE / Math.cos( Constants.DEG2RAD * latitude ) / Constants.EARTH_CIRCUMFERENCE );
+
+}
+
+class CameraSync {
+
+	constructor( map, camera, world ) {
+
+		this.map = map;
+		this.camera = camera;
+		this.active = true;
+		// We're in charge of the camera now!
+		this.camera.matrixAutoUpdate = false;
+		// Postion and configure the world group so we can scale it appropriately when the camera zooms
+		this.world = world || new THREE.Group();
+		this.world.position.x = this.world.position.y =
+    MapConstants.WORLD_SIZE / 2;
+		this.world.matrixAutoUpdate = false;
+		// set up basic camera state
+		this.state = {
+			translateCenter: new THREE.Matrix4().makeTranslation(
+				MapConstants.WORLD_SIZE / 2,
+				- MapConstants.WORLD_SIZE / 2,
+				0
+			),
+			worldSizeRatio: MapConstants.TILE_SIZE / MapConstants.WORLD_SIZE,
+			worldSize: MapConstants.TILE_SIZE * this.map.transform.scale,
+		};
+		// Listen for move events from the map and update the Three.js camera
+		const _this = this;
+		this.map
+			.on( 'move', function () {
+
+				_this.updateCamera();
+
+			} )
+			.on( 'resize', function () {
+
+				_this.setupCamera();
+
+			} );
+
+		this.setupCamera();
+
+	}
+
+	setupCamera() {
+
+		const t = this.map.transform;
+		// if aspect is not reset raycast will fail on map resize
+		this.camera.aspect = t.width / t.height;
+		this.halfFov = t._fov / 2;
+		this.cameraToCenterDistance = ( 0.5 / Math.tan( this.halfFov ) ) * t.height;
+		const maxPitch = ( t._maxPitch * Math.PI ) / 180;
+		this.acuteAngle = Math.PI / 2 - maxPitch;
+		this.updateCamera();
+
+	}
+
+	updateCamera( ev ) {
+
+		if ( ! this.camera ) {
+
+			console.log( 'nocamera' );
+			return;
+
+		}
+
+		const t = this.map.transform;
+
+		this.camera.aspect = t.width / t.height;
+		let farZ = 0;
+		let furthestDistance = 0;
+		this.halfFov = t._fov / 2;
+		const groundAngle = Math.PI / 2 + t._pitch;
+		// pitch seems to influence heavily the depth calculation and cannot be more than 60 = PI/3 < v1 and 85 > v2
+		const pitchAngle = Math.cos( Math.PI / 2 - t._pitch );
+		this.cameraToCenterDistance = ( 0.5 / Math.tan( this.halfFov ) ) * t.height;
+
+		let pixelsPerMeter = 1;
+		const worldSize = this.worldSize();
+
+		pixelsPerMeter = this.mercatorZfromAltitude( 1, t.center.lat ) * worldSize;
+		const fovAboveCenter = t._fov * ( 0.5 + t.centerOffset.y / t.height );
+
+		// Adjust distance to MSL by the minimum possible elevation visible on screen,
+		// this way the far plane is pushed further in the case of negative elevation.
+		const minElevationInPixels = t.elevation
+			? t.elevation.getMinElevationBelowMSL() * pixelsPerMeter
+			: 0;
```

Comment:
How can we compute the elevation in pixels? This is a non constant value since we're using a perspective camera. 1 meter far from the camera will be represented by fewer pixels than a meter closer to the camera.

Reply:
Our purpose is to calculate farZ, which all happens before projection

Reply:
I don't understand. How can you calculate a distance in pixels if it's not relative to the screen?

Reply:
I can't seem to answer your question. These codes refer to the source code of mapboxgl. I have modified this place now. farZ directly uses the one provided by mapboxgl.

Reply:
> pixels

The pixels here are the sizes under the Mercator projection, which is just a possible value calculated.